File: packages/komodo_defi_rpc_methods/lib/src/util/methods_index_generator.dart
import 'dart:io';
import 'package:path/path.dart' as path;

void main(List<String> arguments) {
  if (arguments.isEmpty) {
    print('Please provide the directory path as an argument.');
    exit(1);
  }

  final directory = Directory(arguments.first);
  if (!directory.existsSync()) {
    print('The provided directory does not exist.');
    exit(1);
  }

  final requestClasses = <String, Map<String, dynamic>>{};

  // Traverse the directory and collect all Dart files
  directory.listSync(recursive: true).forEach((file) {
    if (file is File && file.path.endsWith('.dart')) {
      final content = file.readAsStringSync();

      // Regex to match classes that extend BaseRequest
      final classRegex =
          RegExp(r'class\s+(\w+Request)\s+extends\s+BaseRequest');
      final constructorRegex = RegExp(r'(\w+Request)\(([^)]*)\)');

      for (final classMatch in classRegex.allMatches(content)) {
        final className = classMatch.group(1)!;

        final constructorMatch = constructorRegex.firstMatch(content);
        if (constructorMatch != null) {
          final constructorParameters = constructorMatch.group(2)!;
          final methodMatch =
              RegExp(r"method:\s*'([^']*)'").firstMatch(content);

          if (methodMatch != null) {
            final method = methodMatch.group(1)!;
            final methodSegments = method.split('::');

            // Store the class info in the requestClasses map
            Map<String, dynamic> current = requestClasses;

            for (final segment in methodSegments) {
              current = (current.putIfAbsent(
                segment,
                () => {'classes': {}, 'children': {}},
              )['children'] as Map)
                  .cast<String, dynamic>();
            }

            current['classes'] ??= <String, dynamic>{};

            // Store the request class details in the leaf node
            current['classes'][className] = {
              'constructorParameters': constructorParameters,
              'className': className,
            };
          }
        }
      }
    }
  });

  // Check if we found any request classes
  if (requestClasses.isEmpty) {
    print('No request classes were found.');
    exit(1);
  }

  // Generate the index class with nested classes and methods
  final buffer = StringBuffer();
  buffer.writeln('// Auto-generated RPC methods index');
  buffer.writeln('// ignore_for_file: unused_field, unused_element');
  buffer.writeln();
  buffer.writeln(
    "import 'package:komodo_defi_rpc_methods/komodo_defi_rpc_methods.dart';",
  );
  buffer.writeln();
  buffer.writeln('abstract class RpcMethods {');

  void generateNestedClasses(
    Map<String, dynamic> currentMap,
    String indent,
    bool isRoot,
  ) {
    currentMap.forEach((segment, data) {
      final children = (data['children'] as Map? ?? {}).cast<String, dynamic>();
      final classes = (data['classes'] as Map? ?? {}).cast<String, dynamic>();

      final className = '_${segment.pascalCase}Methods';
      if (isRoot) {
        buffer.writeln('$indent// ignore: library_private_types_in_public_api');
        buffer.writeln(
          '${indent}static const $className ${segment.camelCase} = $className();',
        );
      } else {
        buffer.writeln('${indent}class $className {');
        buffer.writeln('$indent  const $className();');
        buffer.writeln();
      }

      classes.forEach((requestClassName, info) {
        final constructorParams = info['constructorParameters'] as String;
        final formattedParams = formatParameters(constructorParams);
        final positionalParams = extractPositionalParams(constructorParams);

        buffer.writeln(
          '$indent  $requestClassName ${requestClassName.camelCase}({$formattedParams}) =>',
        );
        buffer.writeln('$indent      $requestClassName($positionalParams);');
        buffer.writeln();
      });

      if (children.isNotEmpty) {
        generateNestedClasses(children, '$indent  ', false);
      }

      if (!isRoot) {
        buffer.writeln('$indent}');
        buffer.writeln();
      }
    });
  }

  generateNestedClasses(requestClasses, '  ', true);

  buffer.writeln('}');

  final outputPath = path.join(directory.path, 'rpc_methods_index.dart');
  File(outputPath).writeAsStringSync(buffer.toString());

  print('RPC methods index generated at $outputPath');
}

String formatParameters(String params) {
  return params
      .split(',')
      .map((p) => p.trim())
      .where((p) => p.isNotEmpty)
      .map((p) => p.startsWith('required') ? p : 'required $p')
      .join(', ');
}

String extractPositionalParams(String params) {
  return params
      .split(',')
      .map((p) => p.trim())
      .where((p) => p.isNotEmpty)
      .map((p) => p.split(' ').last)
      .join(', ');
}

extension StringExtension on String {
  String get camelCase {
    return this[0].toLowerCase() + substring(1);
  }

  String get pascalCase {
    return this[0].toUpperCase() + substring(1);
  }
}

File: packages/komodo_defi_rpc_methods/lib/src/util/methods_index_generator.py
import os
import re
import sys


def parse_dart_file(file_path):
    with open(file_path, "r") as file:
        content = file.read()

    # Regex to match classes that extend BaseRequest and their constructors
    class_regex = re.compile(r"class\s+(\w+Request)\s+extends\s+BaseRequest")
    constructor_regex = re.compile(r"(\w+Request)\(([^)]*)\)")

    method_data = {}

    # Find the class that extends BaseRequest
    class_match = class_regex.search(content)
    if class_match:
        class_name = class_match.group(1)

        # Find the constructor
        constructor_match = constructor_regex.search(content)
        if constructor_match:
            constructor_params = constructor_match.group(2)

            # Extract the method information
            method_match = re.search(r"method:\s*'([^']*)'", content)
            if method_match:
                method_name = method_match.group(1)
                method_data[class_name] = {
                    "constructorParams": constructor_params,
                    "methodName": method_name,
                }

    return method_data


def generate_dart_code_refined(methods):
    buffer = []
    buffer.append("// Auto-generated RPC methods index")
    buffer.append("abstract class RpcMethods {")

    class_hierarchy = {}

    for class_name, method_info in methods.items():
        method_segments = method_info["methodName"].split("::")
        current_level = class_hierarchy
        for segment in method_segments[:-1]:
            current_level = current_level.setdefault(segment, {})
        current_level[method_segments[-1]] = {
            "className": class_name,
            "constructorParams": method_info["constructorParams"],
        }

    def generate_class_content(buffer, class_name, class_content, level=1):
        indent = "  " * level
        nested_classes = {
            k: v
            for k, v in class_content.items()
            if isinstance(v, dict) and "className" not in v
        }
        methods = {k: v for k, v in class_content.items() if "className" in v}

        buffer.append(f"class _{class_name.capitalize()}Methods {{")
        buffer.append(f"{indent}const _{class_name.capitalize()}Methods();")

        for method_name, method_details in methods.items():
            constructor_params = (
                method_details["constructorParams"].replace("{", "").replace("}", "")
            )
            positional_params = ", ".join(
                [p.split(" ")[-1] for p in constructor_params.split(",") if p.strip()]
            )
            buffer.append(
                f'{indent}{method_details["className"]} {method_name}({constructor_params}) => {method_details["className"]}({positional_params});'
            )

        for nested_class_name, nested_class_content in nested_classes.items():
            buffer.append(
                f"{indent}static const _{nested_class_name.capitalize()}Methods {nested_class_name} = _{nested_class_name.capitalize()}Methods();"
            )
            generate_class_content(
                buffer, nested_class_name, nested_class_content, level + 1
            )

        buffer.append("}")

    for top_level_class_name, top_level_class_content in class_hierarchy.items():
        buffer.append(
            f"  static const _{top_level_class_name.capitalize()}Methods {top_level_class_name} = _{top_level_class_name.capitalize()}Methods();"
        )
        generate_class_content(buffer, top_level_class_name, top_level_class_content)

    buffer.append("}")
    return "\n".join(buffer)


def run_conversion(input_dir, output_dart_file):
    if not os.path.isdir(input_dir):
        print(f"Error: {input_dir} is not a directory.")
        sys.exit(1)

    dart_files = []
    for root, _, files in os.walk(input_dir):
        for file in files:
            if file.endswith(".dart"):
                dart_files.append(os.path.join(root, file))

    if not dart_files:
        print(f"No Dart files found in {input_dir}")
        sys.exit(1)

    parsed_methods = {}
    for dart_file in dart_files:
        method_info = parse_dart_file(dart_file)
        parsed_methods.update(method_info)

    if not parsed_methods:
        print("No methods found extending `BaseRequest`.")
        sys.exit(1)

    dart_code = generate_dart_code_refined(parsed_methods)

    with open(output_dart_file, "w") as file:
        file.write(dart_code)

    print(f"Generated Dart file saved to: {output_dart_file}")


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print(
            "Usage: python generate_rpc_methods.py <input_directory> <output_dart_file>"
        )
        sys.exit(1)

    input_dir = sys.argv[1]
    output_dart_file = sys.argv[2]

    run_conversion(input_dir, output_dart_file)
